### PR TITLE
Configure WebInvocationPrivilegeEvaluator for multiple `SecurityFilterChain`s

### PR DESCRIPTION
**Pull request by marcusdacoregio**
Closes gh-10554, gh-10590

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java [hunk 1]
```diff
@@ -306,6 +315,28 @@ protected Filter performBuild() throws Exception {
...
+			for (Filter filter : securityFilterChain.getFilters()) {
+				if (filter instanceof FilterSecurityInterceptor fsi) {
+					builder.withFilterSecurityInterceptor(fsi, this.servletContext);
+					continue;
+				}
+				if (filter instanceof AuthorizationFilter af) {
+					builder.withAuthorizationManager(af.getAuthorizationManager());
+				}
+			}
+			delegates.add(builder.build());
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java:335]: Do not add if there is no delegates because we don't want to have to invoke match unnecessarily.

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java [hunk 2]
```diff
@@ -284,6 +292,7 @@ protected Filter performBuild() throws Exception {
 		for (SecurityBuilder<? extends SecurityFilterChain> securityFilterChainBuilder : this.securityFilterChainBuilders) {
 			securityFilterChains.add(securityFilterChainBuilder.build());
 		}
+		buildPrivilegeEvaluator(securityFilterChains);
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java:295]: Why is this here the default instead of replacing the default for [`getPrivilegeEvaluator()`](https://github.com/spring-projects/spring-security/blob/a9d98e05f0471dbb888a48a60729279e053649d8/config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java#L252). Putting it in the getter ensures that if anyone tries to obtain it the defaulting is consistent.

File: web/src/main/java/org/springframework/security/web/access/AuthorizationManagerWebInvocationPrivilegeEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	@Override
+	public boolean isAllowed(String uri, Authentication authentication) {
+		return isAllowed(null, uri, null, authentication);
+	}
+
+	@Override
+	public boolean isAllowed(String contextPath, String uri, String method, Authentication authentication) {
+		FilterInvocation filterInvocation = new FilterInvocation(contextPath, uri, method);
+		try {
+			this.authorizationManager.verify(() -> authentication, filterInvocation.getHttpRequest());
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/AuthorizationManagerWebInvocationPrivilegeEvaluator.java:53]: Invoking verify is somewhat costly because it throws an Exception. Instead, the code should invoke the check method.

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		return null;
+	}
+
+	public static class DelegatePrivilegeEvaluator {
+
+		private final Function<HttpServletRequest, Boolean> requestMatcher;
+
+		private final List<WebInvocationPrivilegeEvaluator> privilegeEvaluators;
+
+		public DelegatePrivilegeEvaluator(Function<HttpServletRequest, Boolean> requestMatcher) {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:127]: The code should not allow one to be created without any WebInvocationProivilegeEvaluator instances because we don't want the `RequestMatcher` to be invoked unnecessarily (it decreases the performance every time a check is performed.

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+		}
+
+		public DelegatePrivilegeEvaluator(Function<HttpServletRequest, Boolean> requestMatcher,
+				List<WebInvocationPrivilegeEvaluator> privilegeEvaluators) {
+			Assert.notNull(requestMatcher, "requestMatcher cannot be null");
+			this.requestMatcher = requestMatcher;
+			this.privilegeEvaluators = privilegeEvaluators;
+		}
+
+		public boolean matches(HttpServletRequest request) {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:138]: Perhaps change this to return the `RequestMatcher`.

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+			return this.requestMatcher.apply(request);
+		}
+
+		public List<WebInvocationPrivilegeEvaluator> getPrivilegeEvaluators() {
+			return this.privilegeEvaluators;
+		}
+
+	}
+
+	public static final class Builder {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:148]: Perhaps remove this Builder for now since all of the operations are happening on the DelegatePrivilegeEvaluator anyway (i.e. an invalid DelegatePrivilegeEvaluator can be created).

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+			}
+		}
+		return null;
+	}
+
+	public static class DelegatePrivilegeEvaluator {
+
+		private final Function<HttpServletRequest, Boolean> requestMatcher;
+
+		private final List<WebInvocationPrivilegeEvaluator> privilegeEvaluators;
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:125]: This should be immutable list

File: config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java [hunk 8]
```diff
@@ -306,6 +315,28 @@ protected Filter performBuild() throws Exception {
...
+					builder.withFilterSecurityInterceptor(fsi, this.servletContext);
+					continue;
+				}
+				if (filter instanceof AuthorizationFilter af) {
+					builder.withAuthorizationManager(af.getAuthorizationManager());
+				}
+			}
+			delegates.add(builder.build());
+		}
+		this.privilegeEvaluator = new RequestMatcherDelegatingWebInvocationPrivilegeEvaluator(delegates);
```
Comment (rwinch) [inline config/src/main/java/org/springframework/security/config/annotation/web/builders/WebSecurity.java:337]: instead of initializing privilegeEvaluator it should create an instance and be assigned in getPrivilegeEvaluator. See https://github.com/spring-projects/spring-security/pull/10575/files#diff-2a3768a3ea186b6702859073cc35bc02aa6c8ebf6b20f19a67323f7c20d3bd38R295

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	private DelegatePrivilegeEvaluator getDelegate(String contextPath, String uri, String method) {
+		FilterInvocation filterInvocation = new FilterInvocation(contextPath, uri, method);
+		for (DelegatePrivilegeEvaluator delegate : this.delegates) {
+			if (delegate.matches(filterInvocation.getHttpRequest())) {
+				return delegate;
+			}
+		}
+		return null;
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:118]: Instead of null, change the return type to `List<WebInvocationPrivilegeEvaluator>`. If nothing is found return an empty `List`.

**Review by rwinch — CHANGES_REQUESTED** [review]
Thank you for the PR. I've provided feedback.

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	public static class DelegatePrivilegeEvaluator {
+
+		private final Function<HttpServletRequest, Boolean> requestMatcher;
```
[the remark this answers is not in the document]
Comment (marcusdacoregio) [reply web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:120]: Since we don't have the `RequestMatcher` available when constructing it, we will stick to the `Function` for now

**Comment by marcusdacoregio** [comment]
Thanks for the review @rwinch. I've updated the PR based on your suggestions.

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+	public static class RequestMatcherPrivilegeEvaluator {
+
+		private final Function<HttpServletRequest, Boolean> requestMatcher;
+
+		private final WebInvocationPrivilegeEvaluator privilegeEvaluator;
+
+		public RequestMatcherPrivilegeEvaluator(Function<HttpServletRequest, Boolean> requestMatcher,
+				WebInvocationPrivilegeEvaluator privilegeEvaluator) {
+			Assert.notNull(requestMatcher, "requestMatcher cannot be null");
```
[the remark this answers is not in the document]
Comment (rwinch) [reply web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:126]: It looks like there is still the need for an empty check on privilegeEvalutators

**Review by rwinch — CHANGES_REQUESTED** [review]

File: web/src/main/java/org/springframework/security/web/access/AuthorizationManagerWebInvocationPrivilegeEvaluator.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.util.Assert;
+
+/**
+ * An implementation of {@link WebInvocationPrivilegeEvaluator} which delegates the checks
+ * to an instance of {@link AuthorizationManager}
+ *
+ * @author Marcus Da Coregio
+ * @since 5.7
+ */
+public final class AuthorizationManagerWebInvocationPrivilegeEvaluator implements WebInvocationPrivilegeEvaluator {
```
[the remark this answers is not in the document]
Comment (marcusdacoregio) [reply web/src/main/java/org/springframework/security/web/access/AuthorizationManagerWebInvocationPrivilegeEvaluator.java:34]: I've splitted it in another commit that closes gh-10590

File: web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcusdacoregio) [reply web/src/main/java/org/springframework/security/web/access/RequestMatcherDelegatingWebInvocationPrivilegeEvaluator.java:126]: There will be some scenarios where it makes sense to have an empty list. When using `web.ignoring(...)` we have a `SecurityFilterChain` with no filters, therefore no `WebInvocationPrivilegeEvaluator`s. So, in this scenario we can accept an empty list

**Review by rwinch — APPROVED** [review]
Thanks for the PR! LGTM
